src/log-store/src/raft_engine/backend.rs

```diff
@@ -0,0 +1,548 @@
+        self
+    }
+
+    async fn range(&self, req: RangeRequest) -> Result<RangeResponse, Self::Error> {
+        let mut res = vec![];
+        self.engine
+            .read()
+            .unwrap()
+            .scan_raw_messages(
+                SYSTEM_NAMESPACE,
+                Some(&req.key),
+                Some(&req.range_end),
+                false,
+                |key, value| {
+                    res.push(KeyValue {
+                        key: key.to_vec(),
+                        value: value.to_vec(),
+                    });
+                    true
+                },
+            )
+            .context(RaftEngineSnafu)?;
+        Ok(RangeResponse {
+            kvs: res,
+            more: false,
+        })
+    }
+
+    async fn put(&self, req: PutRequest) -> Result<PutResponse, Self::Error> {
+        let PutRequest {
+            key,
+            value,
+            prev_kv,
+        } = req;
+
+        let mut prev = None;
+        let engine = self.engine.read().unwrap();
+        if prev_kv {
+            prev = engine_get(&engine, &key)?;
+        }
+        engine_put(&engine, key, value)?;
+        Ok(PutResponse { prev_kv: prev })
+    }
+
+    async fn batch_put(&self, req: BatchPutRequest) -> Result<BatchPutResponse, Self::Error> {
+        let BatchPutRequest { kvs, prev_kv } = req;
+        let mut batch = LogBatch::with_capacity(kvs.len());
+
+        let mut prev_kvs = if prev_kv {
+            Vec::with_capacity(kvs.len())
+        } else {
+            vec![]
+        };
+
+        for kv in kvs {
+            if prev_kv && let Some(kv) = engine_get(&self.engine.read().unwrap(), &kv.key)? {
+                prev_kvs.push(kv);
+            }
+            batch
+                .put(SYSTEM_NAMESPACE, kv.key, kv.value)
+                .context(RaftEngineSnafu)?;
+        }
+
+        self.engine
+            .read()
+            .unwrap()
+            .write(&mut batch, false)
+            .context(RaftEngineSnafu)?;
+
+        Ok(BatchPutResponse { prev_kvs })
+    }
+
+    async fn batch_get(&self, req: BatchGetRequest) -> Result<BatchGetResponse, Self::Error> {
+        let mut response = BatchGetResponse {
+            kvs: Vec::with_capacity(req.keys.len()),
+        };
+        let engine = self.engine.read().unwrap();
+        for key in req.keys {
+            let value = engine.get(SYSTEM_NAMESPACE, &key).unwrap();
+            response.kvs.push(KeyValue { key, value });
+        }
+        Ok(response)
+    }
+
+    async fn compare_and_put(
+        &self,
+        req: CompareAndPutRequest,
+    ) -> Result<CompareAndPutResponse, Self::Error> {
+        let CompareAndPutRequest { key, expect, value } = req;
+
+        let mut batch = LogBatch::with_capacity(1);
+        let engine = self.engine.write().unwrap();
+        let existing = engine_get(&engine, &key)?;
+        let eq = existing
+            .as_ref()
+            .map(|kv| kv.value == expect)
+            .unwrap_or(false);
```

Comment:
I find that in `MemoryKvBackend`, if the key does not exist while the `expect` is `vec![]`, it's considered "equal". I remember this behavior was there before I refactored the `MemoryKvBackend`. Is it a design choice? @fengjiachun

Reply:
fixed in [257c7a6](https://github.com/GreptimeTeam/greptimedb/pull/2280/commits/257c7a6d3cd42ac501dfecfa3533502c5011f0e9#diff-74b285f40ce368ca6fe6df02183ca1e98c68112554e907aa9e7271ee6ca2d6feR216-R220)